File: frontend/src/lib/basePriceLabels.ts
import type { FlatBox } from "../types";

export const BASE_PRICE_LABELS = {
  base: "Nền giá",
  breakout: "Phá vỡ nền giá",
  breakUp: "Nổ hướng lên",
  breakDown: "Gãy nền",
} as const;

export const BREAK_UP_MIN_PERCENT = 3;

export function resolveBasePriceEventLabel(
  flatBox: FlatBox,
  latestPrice: number,
): string {
  if (flatBox.eventLabel) return flatBox.eventLabel;

  if (latestPrice < flatBox.boxLow) return BASE_PRICE_LABELS.breakDown;

  if (
    flatBox.isBreakoutConfirmed &&
    (flatBox.priceGainPercent ?? 0) >= BREAK_UP_MIN_PERCENT
  ) {
    return BASE_PRICE_LABELS.breakUp;
  }

  if (flatBox.isBreakoutConfirmed) return BASE_PRICE_LABELS.breakout;

  return BASE_PRICE_LABELS.base;
}

export function flatBoxCardSubtitle(flatBox: FlatBox, latestPrice: number): string {
  if (!flatBox.isBreakoutConfirmed) {
    return `Đang tích lũy — ${flatBox.refBoxPeriod}`;
  }
  return resolveBasePriceEventLabel(flatBox, latestPrice);
}

File: mobile/lib/core/labels/base_price_labels.dart
class BasePriceLabels {
  static const String base = 'Nền giá';
  static const String breakout = 'Phá vỡ nền giá';
  static const String breakUp = 'Nổ hướng lên';
  static const String breakDown = 'Gãy nền';
  static const double breakUpMinPercent = 3;

  static String resolveEventLabel(
    Map<String, dynamic> box,
    double latestPrice,
  ) {
    final eventLabel = box['eventLabel'] as String?;
    if (eventLabel != null && eventLabel.isNotEmpty) return eventLabel;

    final boxLow = (box['boxLow'] as num?)?.toDouble() ?? 0;
    final confirmed = box['isBreakoutConfirmed'] as bool? ?? false;
    final priceGain = (box['priceGainPercent'] as num?)?.toDouble() ?? 0;

    if (latestPrice < boxLow) return breakDown;
    if (confirmed && priceGain >= breakUpMinPercent) return breakUp;
    if (confirmed) return breakout;
    return base;
  }

  static String cardSubtitle(Map<String, dynamic> box, double latestPrice) {
    final confirmed = box['isBreakoutConfirmed'] as bool? ?? false;
    if (!confirmed) {
      final refPeriod = box['refBoxPeriod'] as String? ?? '';
      return 'Đang tích lũy — $refPeriod';
    }
    return resolveEventLabel(box, latestPrice);
  }
}
